[mfl] Add new port

cpp-niel opened the pull request:
**Add port for the mfl library**

- What does your PR fix?
  * Fixes #16343

- Which triplets are supported/not supported? Have you updated the CI baseline?
  * Dynamic x86 and x64 windows triplets are not supported

- Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?
  * Yes (as far as I am aware)

cpp-niel committed:
Add control and portfile for mfl port

cpp-niel committed:
Add mfl port

cpp-niel committed:
Merge branch 'add_mfl_port' of https://github.com/cpp-niel/vcpkg into add_mfl_port

cpp-niel committed:
reference version better header inclusion

cpp-niel committed:
reference version with fmt linkage

NancyLi1013 commented on ports/mfl/portfile.cmake:
```diff
@@ -0,0 +1,18 @@
+
+vcpkg_install_cmake()
+
+vcpkg_fixup_cmake_targets(CONFIG_PATH lib/cmake/mfl TARGET_PATH share/mfl)
```
```suggestion
vcpkg_fixup_cmake_targets(CONFIG_PATH lib/cmake/mfl)
```
It's unnecessary to specify `TARGET_PATH` here, since it will point to `share/${PORT}` by default. Please see https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/vcpkg_fixup_cmake_targets.md.

NancyLi1013 commented on ports/mfl/portfile.cmake:
```diff
@@ -0,0 +1,18 @@
+
+vcpkg_fixup_cmake_targets(CONFIG_PATH lib/cmake/mfl TARGET_PATH share/mfl)
+
+file(INSTALL ${SOURCE_PATH}/LICENSE DESTINATION ${CURRENT_PACKAGES_DIR}/share/mfl RENAME copyright)
```
```suggestion
file(INSTALL ${SOURCE_PATH}/LICENSE DESTINATION ${CURRENT_PACKAGES_DIR}/share/${PORT} RENAME copyright)
```

NancyLi1013 commented:
The failures on x86-windows:
```
D:\installed\x86-windows\include\range/v3/algorithm/find.hpp(53): error C2220: the following warning is treated as an error
D:\installed\x86-windows\include\range/v3/algorithm/find.hpp(53): warning C4389: '==': signed/unsigned mismatch
```

The failures on x64-windows:

```
FAILED: tests/unit_tests/unit_tests.exe 
cmd.exe /C "cd . && D:\downloads\tools\cmake-3.19.2-windows\cmake-3.19.2-win32-x86\bin\cmake.exe -E vs_link_exe --intdir=tests\unit_tests\CMakeFiles\unit_tests.dir --rc="C:\PROGRA~2\Windows Kits\10\bin\10.0.19041.0\x64\rc.exe" --mt="C:\PROGRA~2\Windows Kits\10\bin\10.0.19041.0\x64\mt.exe" --manifests  -- "C:\PROGRA~2\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.28.29333\bin\Hostx64\x64\link.exe"  tests\unit_tests\CMakeFiles\unit_tests.dir\dist.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\framework\mock_font_face.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\layout.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\main.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\noad\accent.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\noad\big_op.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\noad\fraction.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\noad\left_right.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\noad\math_char.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\noad\noad.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\noad\overline.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\noad\radical.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\noad\script.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\noad\underline.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\node\box.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\node\glue.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\node\hlist.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\node\node.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\node\vlist.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\node\vstack.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\parser\accent.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\parser\big_op.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\parser\font_group.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\parser\fraction.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\parser\function.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\parser\left_right.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\parser\lexer.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\parser\line.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\parser\math_char.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\parser\math_space.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\parser\parse.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\parser\radical.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\parser\script.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\parser\tokens.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\parser\unicode_index.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\units.cpp.obj  /out:tests\unit_tests\unit_tests.exe /implib:tests\unit_tests\unit_tests.lib /pdb:tests\unit_tests\unit_tests.pdb /version:0.0 /machine:x64 /nologo    /debug /INCREMENTAL /subsystem:console  src\mfl.lib  D:\installed\x64-windows\debug\lib\fmtd.lib  kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib && cd ."
LINK Pass 1: command "C:\PROGRA~2\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.28.29333\bin\Hostx64\x64\link.exe tests\unit_tests\CMakeFiles\unit_tests.dir\dist.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\framework\mock_font_face.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\layout.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\main.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\noad\accent.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\noad\big_op.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\noad\fraction.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\noad\left_right.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\noad\math_char.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\noad\noad.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\noad\overline.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\noad\radical.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\noad\script.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\noad\underline.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\node\box.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\node\glue.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\node\hlist.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\node\node.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\node\vlist.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\node\vstack.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\parser\accent.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\parser\big_op.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\parser\font_group.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\parser\fraction.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\parser\function.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\parser\left_right.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\parser\lexer.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\parser\line.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\parser\math_char.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\parser\math_space.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\parser\parse.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\parser\radical.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\parser\script.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\parser\tokens.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\parser\unicode_index.cpp.obj tests\unit_tests\CMakeFiles\unit_tests.dir\units.cpp.obj /out:tests\unit_tests\unit_tests.exe /implib:tests\unit_tests\unit_tests.lib /pdb:tests\unit_tests\unit_tests.pdb /version:0.0 /machine:x64 /nologo /debug /INCREMENTAL /subsystem:console src\mfl.lib D:\installed\x64-windows\debug\lib\fmtd.lib kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib /MANIFEST /MANIFESTFILE:tests\unit_tests\CMakeFiles\unit_tests.dir/intermediate.manifest tests\unit_tests\CMakeFiles\unit_tests.dir/manifest.res" failed (exit code 1104) with the following output:
LINK : fatal error LNK1104: cannot open file 'src\mfl.lib'
ninja: build stopped: subcommand failed.
```

The failures on x64-linux:

```
/mnt/vcpkg-ci/buildtrees/mfl/src/v0.0.1-fdd809ab97.clean/include/mfl/detail/quantity.hpp:5:10: fatal error: compare: No such file or directory
 #include <compare>
          ^~~~~~~~~
compilation terminated.
```

You can get the log here https://dev.azure.com/vcpkg/public/_build/results?buildId=49677&view=artifacts&pathAsName=false&type=publishedArtifacts

cpp-niel committed:
Apply suggestions from code review

Co-authored-by: NancyLi1013 <46708020+NancyLi1013@users.noreply.github.com>

cpp-niel committed:
Set ports that don't build to fail in the ci baseline

cpp-niel committed:
Merge branch 'add_mfl_port' of https://github.com/cpp-niel/vcpkg into add_mfl_port

cpp-niel committed:
add version

NancyLi1013 commented on scripts/ci.baseline.txt:
```diff
@@ -886,6 +886,9 @@ metrohash:arm-uwp=fail
 metrohash:arm64-windows=fail
+mfl:x64-linux=fail
+mfl:x86-windows=fail
+mfl:x64-windows=fail
```
Does this port not support these triplets?

cpp-niel replied:
It doesn't support the dynamic Windows triplets - only static builds on Windows.

Theoretically, it does support x64-linux, but it requires gcc 10. The CI build appears to be using gcc 7.5. If there is a better way of dealing with this I would be really grateful if you could point me in the right direction.

NancyLi1013 replied:
For `x86-windows` and `x64-windows`, you can try to add the following codes to portfile.cmake.
```
if(VCPKG_TARGET_IS_WINDOWS) 
    vcpkg_check_linkage(ONLY_STATIC_LIBRARY)
endif()
```
For x64-linux, since most users use gcc below 10. So keep it work normally for the specific users, we may set it as `skip `in ci.baseline.txt. It would be better to add the related message to portfile.cmake.

cpp-niel replied:
Thanks so much for your help. I have added the linkage check and the message to portfile.cmake and I have set the dynamic windows builds to `skip` in ci.baseline.txt.

NancyLi1013 replied:
You can try to add `supports: static` to check if CI can test passed on` x86-windows` and `x64-windows` triplets.

If so, it's unnecessary to add them to fail list in ci.baseline.txt.

cpp-niel replied:
I've made this change and removed the `fail` lines from ci.baseline.txt. Everything appears to be working. Thank you.

cpp-niel committed:
Check for static linkage and add message about C++20 requirements.

cpp-niel committed:
add fail to ci baseline for dynamic windows builds

cpp-niel committed:
version update

NancyLi1013 commented on ports/mfl/vcpkg.json:
```diff
@@ -0,0 +1,15 @@
+{
+  "name": "mfl",
+  "version-string": "0.0.1",
```
For package versions, please refer to this doc https://github.com/microsoft/vcpkg/blob/master/docs/specifications/versioning.md#22-package-versions to check if it is correct.

cpp-niel replied:
Set to `version` rather than `version-string`

NancyLi1013 commented on ports/mfl/portfile.cmake:
```diff
@@ -0,0 +1,24 @@
+    vcpkg_check_linkage(ONLY_STATIC_LIBRARY)
+endif()
+
+message("Warning: `mfl` is a C++20 library and requires a corresponding compiler. GCC 10, Clang 10 and MSVC 2019 16.8 are known to work.")
```
```suggestion
message(WARNING "${PORT} is a C++20 library and requires a corresponding compiler. GCC 10, Clang 10 and MSVC 2019 16.8 are known to work.")
```

cpp-niel replied:
Adopted suggested change

cpp-niel committed:
Code review improvements

- Improve warning message that the port requires a C++20 compiler
- Use `version` rather than `version-string` in manifest
- Use `"supports": "static"` in manifest instead of setting dynamic
  builds to `fail` in ci baseline.

cpp-niel committed:
update version

NancyLi1013 commented:
LGTM now, thanks for adding this new port @cpp-niel.

cpp-niel commented:
@NancyLi1013, thanks for your help.

cpp-niel committed:
Merge branch 'master' into add_mfl_port

NancyLi1013 commented:
/azp run
